core/src/main/java/hudson/link/ConsoleURLProvider.java

```diff
@@ -0,0 +1,59 @@
+package hudson.link;
+
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.Util;
+import hudson.model.Run;
+import jenkins.model.Jenkins;
+
+/**
+ * Extension point for providing console urls
+ * @since TODO
+ */
+public abstract class ConsoleURLProvider implements ExtensionPoint {
+
+    /**
+     * Provide the console url of the highest registered ordinal implementation.
+     * Defaults to {@link ConsoleURLProviderImpl} otherwise.
+     * @since TODO
+     */
+    public String getConsoleURL(Run<?, ?> run) {
+        return get().getConsoleURL(run);
+    }
+
+    /**
+     * Retrieve all implementations of ConsoleURLProvider.
+     * @since TODO
+     */
+    public static ExtensionList<ConsoleURLProvider> all() {
+        return ExtensionList.lookup(ConsoleURLProvider.class);
+    }
+
+    /**
+     * Retrieve the highest registered ordinal implementation.
+     * @since 2.417
+     */
+    public static ConsoleURLProvider get() {
+        return all().stream().findFirst().orElse(ConsoleURLProviderImpl.INSTANCE);
+    }
+
+    @Extension(ordinal = -100)
+    public static class ConsoleURLProviderImpl extends ConsoleURLProvider {
+
+        static final ConsoleURLProvider INSTANCE = new ConsoleURLProviderImpl();
+
+        @Override
+        public String getConsoleURL(Run<?, ?> run) {
+            return getRoot() + Util.encode(run.getUrl()) + "console";
+        }
+
+        public String getRoot() {
+            String root = Jenkins.get().getRootUrl();
+            if (root == null) {
+                root = "http://unconfigured-jenkins-location/";
```

Comment:
Is there a legitimate use case for when this code would be reached?

Display URL API has this because the URLs are used outside Jenkins and computed in a non-request context. It seems these URLs are always used in a request context, as they're used to render parts of the UI?

Reply:
Good point, it would probably be better to design the new API to work with URLs relative to the context root, and then use something like `'/' + Stapler.getCurrentRequest().getContextPath() + getConsoleURL(run)` to build the console URLs.

The implementation of the extension that should go into https://github.com/jenkinsci/display-url-api-plugin/pull/194 would need to convert the absolute URLs into context path-relative URLs though, and I guess what I was thinking [here](https://github.com/jenkinsci/pipeline-input-step-plugin/pull/145#discussion_r1281880000) would no longer apply, since we would want to use `ConsoleURLProvider` for all links in the Jenkins UI and `DisplayUrlProvider.getConsoleUrl` only when producing external links.

Reply:
Thinking about it some more -- just after the ping comment, sorry Cyrille -- this could also easily be implemented by having a Jenkins-internal URL that redirects to the outside system, instead of linking there directly.

Reply:
Thanks @daniel-beck. Regarding the OpenTelemetry integrations where Jenkins pipeline logs are persisted in an external logs management system rather than in the Jenkins home, all integrations I am aware of want to use Jenkins's built-in logs visualization UX.
Those integrations